src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -102,6 +89,113 @@ pub struct ConfigSpace {
+    deferred_descriptor: Option<ParsedDescriptorChain>,
+}
+
+impl RxBuffers {
+    /// Create a new [`RxBuffers`] object for storing guest memory for performing RX
+    fn new() -> Result<Self, IoVecError> {
+        Ok(Self {
+            min_buffer_size: 0,
+            iovec: IoVecBufferMut::new()?,
+            parsed_descriptors: VecDeque::with_capacity(FIRECRACKER_MAX_QUEUE_SIZE.into()),
+            deferred_descriptor: None,
+        })
+    }
+
+    /// Add a new `DescriptorChain` that we received from the RX queue in the buffer.
+    ///
+    /// SAFETY: The `DescriptorChain` cannot be referencing the same memory location as any other
+    /// `DescriptorChain`. (See also related comment in
+    /// [`IoVecBufferMut::append_descriptor_chain`]).
+    unsafe fn add_buffer(
+        &mut self,
+        mem: &GuestMemoryMmap,
+        head: DescriptorChain,
+    ) -> Result<(), AddRxBufferError> {
+        let parsed_dc = self.iovec.append_descriptor_chain(mem, head)?;
+        if parsed_dc.length < self.min_buffer_size {
+            self.iovec.drop_descriptor_chain(&parsed_dc);
+            return Err(AddRxBufferError::BufferTooSmall);
+        }
+        self.parsed_descriptors.push_back(parsed_dc);
+        Ok(())
+    }
+
+    /// Returns the number of available `iovec` objects.
+    #[inline(always)]
+    fn len(&self) -> usize {
+        self.iovec.len()
+    }
+
+    /// Mark the first `size` bytes of available memory as used.
+    ///
+    /// The `size` should be the number of bytes previously written in the [`RxBuffers`]. This
+    /// means that `size` cannot be more than the currently available bytes in the underlying
+    /// `IoVecBufferMut`.
+    fn mark_used(&mut self, size: u32, metrics: &NetDeviceMetrics) {
+        // Since we were able to write a frame in guest memory, we should have at least one
+        // descriptor chain here. If not, we have a bug, so fail fast, since the device is
+        // fundamentally broken.
+        let mut parsed_dc = self.parsed_descriptors.pop_front().expect(
+            "net: internal bug. Mismatch between written frame size and available descriptors",
+        );
+
+        // We are trying to mark as used more bytes than a single descriptor. However without
+        // VIRTIO_NET_F_MRG_RXBUF negotiated (we don't support it currently) the guest can't handle
+        // frames spanning multiple DescriptorChains. There is nothing we can do now, just pass the
+        // descriptor back to the guest without claiming any written bytes
+        if size > parsed_dc.length {
+            parsed_dc.length = 0;
+            metrics.rx_fails.inc();
+        } else {
+            parsed_dc.length = size;
+            metrics.rx_bytes_count.add(size as u64);
+            metrics.rx_packets_count.inc();
+        }
+        self.iovec.drop_descriptor_chain(&parsed_dc);
+        self.deferred_descriptor = Some(parsed_dc);
```

Comment:
I would prefer if we would write used descriptor into chain right here, but do actual notification in `finish_frame` (like it is done here https://github.com/firecracker-microvm/firecracker/pull/4813/files#diff-205f89a698c2fafa14dcbd3c76dfee31dd597294c8334bc6b82d8c92814a497fR226)
This way it will require less changes for MRG_RXBUF support and it will reduce the size of the `RxBuffers` struct as we will need to store `u16` instead of `Option<ParsedDescriptorChain>`.

Reply:
I am not sure I understand what you mean here. We need the number of the bytes that were actually written as well, right?

Reply:
You have `size` argument which is the number of bytes written into chain. Currently you do this: `parsed_dc.length = size;` and save the `parsed_dc`. But you can use this `size` argument to write a used descriptor into queue at this point. And in the `finish_frame` just bump queue by 1.

Reply:
Ok, I see now. You want to break down `add_used` in two parts. What do we gain with that?

Reply:
This:

>This way it will require less changes for MRG_RXBUF support and it will reduce the size of the RxBuffers struct as we will need to store u16 instead of Option<ParsedDescriptorChain>.

`MRG_RXBUF` will need to know all chains it used for packet. With current approach we will need `Vec<ParsedDescriptorChain>` to store them. But we can avoid this by splitting `add_used` into 2 parts and only store the number of heads we used in the `RxBuffers`.

Reply:
Well.. in my opinion it is better than current version anyway, but ok, lest's leave this for mrg_rxbuf impl.